### PR TITLE
Data independent form compilation for Python interface

### LINKED ISSUE #3111: Use pure UFL/basix.ufl forms in compilation in DOLFINx Python interface
### Describe new/missing feature

In C++, we use pure ufl/basix python files to compile forms. 
I would like the same user-interface from Python.
This is for instance neat when one uses multiple meshes in the same code, or want to make a library containing various forms, that could be used with either the Python or C++ interface. 
A minimal working product is described below.
Here I use `get_integration_domains`, current a hidden function in DOLFINx python interface to compute the integration entities.
The only change to make this work is to expose `dolfinx::fem::create_form` to the Python interface:
```c++
  m.def(
      pymethod_create_form.c_str(),
      [](std::uintptr_t form,
         const std::vector<
             std::shared_ptr<const dolfinx::fem::FunctionSpace<U>>>& spaces,
         const std::map<std::string,
                        std::shared_ptr<const dolfinx::fem::Function<T, U>>>&
             coefficients,
         const std::map<std::string,
                        std::shared_ptr<const dolfinx::fem::Constant<T>>>&
             constants,
         const std::map<dolfinx::fem::IntegralType,
                        std::vector<std::pair<std::int32_t,
                                              std::span<const std::int32_t>>>>&
             subdomains,
         std::shared_ptr<const dolfinx::mesh::Mesh<U>> mesh = nullptr)
      {
        std::map<
            dolfinx::fem::IntegralType,
            std::vector<std::pair<std::int32_t, std::span<const std::int32_t>>>>
            sd;
        for (auto& [itg, data] : subdomains)
        {
          std::vector<std::pair<std::int32_t, std::span<const std::int32_t>>> x;
          for (auto& [id, idx] : data)
            x.emplace_back(id, std::span(idx.data(), idx.size()));
          sd.insert({itg, std::move(x)});
        }

        ufcx_form* p = reinterpret_cast<ufcx_form*>(form);
        return dolfinx::fem::create_form<T, U>(*p, spaces, coefficients,
                                               constants, sd, mesh);
      },
      nb::arg("form"), nb::arg("spaces"), nb::arg("coefficients"),
      nb::arg("constants"), nb::arg("subdomains"), nb::arg("mesh"),
      "Create Form from a pointer to ufcx_form.");
```
**Things to improve**
- Currently one has to use the `ufl` - coeffs/constants to map to the dolfinx coeffs/constants. This is due to the fact that we cannot supply names to dolfinx.jit.ffcx_jit or ufl, meaning that the coefficients are named `w0, w1,..` and constants `c0, c1, ...`.

### Suggested user interface

```python3
import ufl
import basix.ufl
import numpy as np


from mpi4py import MPI
import dolfinx

import numpy.typing as npt
import typing


from dataclasses import dataclass


@dataclass
class GeneralForm:
    ufl_form: ufl.Form
    ufcx_form: typing.Any
    module: typing.Any
    code: str


def compile_form(form: ufl.Form, comm: MPI.Intracomm) -> GeneralForm:
    compiled_form = dolfinx.jit.ffcx_jit(comm, J)
    return GeneralForm(form, *compiled_form)


def get_integration_domains(integral_type, subdomain):
    """Get integration domains from subdomain data"""
    if subdomain is None:
        return []
    else:
        try:
            if integral_type in (
                dolfinx.fem.IntegralType.exterior_facet,
                dolfinx.fem.IntegralType.interior_facet,
            ):
                tdim = subdomain.topology.dim
                subdomain._cpp_object.topology.create_connectivity(tdim - 1, tdim)
                subdomain._cpp_object.topology.create_connectivity(tdim, tdim - 1)
            domains = dolfinx.cpp.fem.compute_integration_domains(
                integral_type, subdomain._cpp_object
            )
            return [(s[0], np.array(s[1])) for s in domains]
        except AttributeError:
            return [(s[0], np.array(s[1])) for s in subdomain]


def form_cpp_creator(
    dtype: npt.DTypeLike,
) -> typing.Union[
    dolfinx.cpp.fem.Form_float32,
    dolfinx.cpp.fem.Form_float64,
    dolfinx.cpp.fem.Form_complex64,
    dolfinx.cpp.fem.Form_complex128,
]:
    """Return the wrapped C++ class of a variational form of a specific scalar type.

    Args:
        dtype: Scalar type of the required form class.

    Returns:
        Wrapped C++ form class of the requested type.

    Note:
        This function is for advanced usage, typically when writing
        custom kernels using Numba or C.
    """
    if np.issubdtype(dtype, np.float32):
        return dolfinx.cpp.fem.create_form_float32
    elif np.issubdtype(dtype, np.float64):
        return dolfinx.cpp.fem.create_form_float64
    elif np.issubdtype(dtype, np.complex64):
        return dolfinx.cpp.fem.create_form_complex64
    elif np.issubdtype(dtype, np.complex128):
        return dolfinx.cpp.fem.create_form_complex128
    else:
        raise NotImplementedError(f"Type {dtype} not supported.")


def create_form(
    form: GeneralForm,
    mesh: dolfinx.mesh.Mesh,
    coefficient_map: typing.Dict[str, dolfinx.fem.Function],
    constant_map: typing.Dict[str, dolfinx.fem.Constant],
):
    sd = form.ufl_form.subdomain_data()
    (domain,) = list(sd.keys())  # Assuming single domain

    # Subdomain markers (possibly empty list for some integral types)
    subdomains = {
        dolfinx.fem.forms._ufl_to_dolfinx_domain[key]: get_integration_domains(
            dolfinx.fem.forms._ufl_to_dolfinx_domain[key], subdomain_data[0]
        )
        for (key, subdomain_data) in sd.get(domain).items()
    }
    coefficients = {
        f"w{u.count()}": uh._cpp_object for (u, uh) in coefficient_map.items()
    }
    constants = {f"c{c.count()}": ch._cpp_object for (c, ch) in constant_map.items()}
    ftype = form_cpp_creator(dolfinx.default_scalar_type)
    f = ftype(
        form.module.ffi.cast("uintptr_t", form.module.ffi.addressof(form.ufcx_form)),
        [],
        coefficients,
        constants,
        subdomains,
        mesh._cpp_object,
    )
    return dolfinx.fem.Form(f, form.ufcx_form, form.code)


# Create ufl form exclusively with basix.ufl and UFL
c_el = basix.ufl.element("Lagrange", "triangle", 1, shape=(2,))
domain = ufl.Mesh(c_el)
el = basix.ufl.element("Lagrange", "triangle", 2)
V = ufl.FunctionSpace(domain, el)
u = ufl.Coefficient(V)
w = ufl.Coefficient(V)
c = ufl.Constant(domain)
e = ufl.Constant(domain)
J = c * e * u * w * ufl.dx(domain=domain)

# Compile form using dolfinx.jit.ffcx_jit
compiled_form = compile_form(J, MPI.COMM_WORLD)


def create_and_integrate(N, compiled_form):

    mesh = dolfinx.mesh.create_unit_square(MPI.COMM_WORLD, N, N)
    el_2 = basix.ufl.element("Lagrange", "triangle", 2)
    Vh = dolfinx.fem.functionspace(mesh, el_2)
    uh = dolfinx.fem.Function(Vh)
    uh.interpolate(lambda x: x[0])
    wh = dolfinx.fem.Function(Vh)
    wh.interpolate(lambda x: x[1])
    eh = dolfinx.fem.Constant(mesh, 3.0)
    ch = dolfinx.fem.Constant(mesh, 2.0)
    form = create_form(compiled_form, mesh, {u: uh, w: wh}, {c: ch, e: eh})
    print(mesh.comm.allreduce(dolfinx.fem.assemble_scalar(form), op=MPI.SUM))


# Create various meshes, that all uses this compiled form with a map from ufl to dolfinx functions and constants
for i in range(1, 4):
    create_and_integrate(i, compiled_form)
```

### PR DESCRIPTION
**Pull request by jorgensd**
Resolves #3111.

One can now compile a form once without meshes, and then in turn use if for a sequence meshes.

This is practical for mesh refinement studies (error estimation), multigrid schemes and eventually P-adaptive FEM with codim-0 submeshes.

Another nice feature is that one can make libraries that define variational forms, and use them both in C++ and Python.

It would also make explaining the inner workings of FEniCS a lot easier, as one can start by introducing Basix/ufl, define the variational forms, and then go on to discretizing the problem. This came up when I lectured FEniCSx in Sorbonne last year, as there was a clear leap from ufl/basix to ffcx to dolfinx.

### CODE REVIEW

File: python/dolfinx/fem/forms.py [hunk 1]
```diff
@@ -333,3 +360,137 @@ def unique_spaces(V):
...
+
+
+def create_form(
+    form: CompiledForm,
+    mesh: Mesh,
+    coefficient_map: dict[ufl.Function, function.Function],
+    constant_map: dict[ufl.Constant, function.Constant],
+) -> Form:
+    """
+    Create a Form object from a data-independent compilde form
```
Comment (chrisrichardson) [inline python/dolfinx/fem/forms.py:433]:
```suggestion
    Create a Form object from a data-independent compiled form
```

File: python/dolfinx/fem/forms.py [hunk 2]
```diff
@@ -333,3 +360,140 @@ def unique_spaces(V):
...
+    """
+    Compiled UFL form without associated DOLFINx data
+
+    """
+
+    ufl_form: ufl.Form  # The original ufl form
+    ufcx_form: typing.Any  # The compiled form
+    module: typing.Any  #  The module
+    code: str  # The source code
+    dtype: npt.DTypeLike  # data type of coefficients and constants
```
Comment (chrisrichardson) [inline python/dolfinx/fem/forms.py:376]: Is this needed? Surely the form itself already has a `dtype` and the constants and coefficients need to match that.

Comment (jorgensd) [reply python/dolfinx/fem/forms.py:376]: You do not have the coefficients at compilation time (with dtype). You only have the abstract ufl objects.

Comment (chrisrichardson) [reply python/dolfinx/fem/forms.py:376]: OK, but the `ufcx_form` is compiled for a particular `dtype`. Maybe we do need to store it here, but I think it should just be called the "dtype of the form". After all, you might not have any constants or coefficients.

File: python/dolfinx/fem/forms.py [hunk 3]
```diff
@@ -333,3 +360,139 @@ def unique_spaces(V):
...
```
Comment (jorgensd) [inline python/dolfinx/fem/forms.py:376]:
```suggestion
    dtype: npt.DTypeLike  # data type used for the `ufcx_form`
```

**Comment by jorgensd** [comment]
Example of how one can now write up a DG method:

`dg_form.py`
```python
import ufl
from basix.ufl import element

# Create a symbolic representation of a mesh
c_el = element("Lagrange", "triangle", 1, shape=(2,))
domain = ufl.Mesh(c_el)

# Create a symbolic representation of a function space
el = element("Lagrange", "triangle", 3, discontinuous=True)
V = ufl.FunctionSpace(domain, el)

# Define problem specific variables
h = 2 * ufl.Circumradius(domain)
n = ufl.FacetNormal(domain)
x, y = ufl.SpatialCoordinate(domain)
g = ufl.sin(2*ufl.pi*x)+ufl.cos(y)
f = ufl.Coefficient(V)
alpha = ufl.Constant(domain)
gamma = ufl.Constant(domain)
u = ufl.TrialFunction(V)
v = ufl.TestFunction(V)

# Define variational formulation
ds = ufl.Measure("ds", domain=domain)
dx = ufl.Measure("dx", domain=domain)
dS = ufl.Measure("dS", domain=domain)

F = ufl.inner(ufl.grad(u), ufl.grad(v)) * dx - f * v * dx


# # Nitsche terms
def flux_term(u, v):
    return -ufl.dot(n, ufl.grad(u)) * v


F += flux_term(v, u) * ds + alpha / h * u * v * ds + flux_term(u, v) * ds
F -= flux_term(v, g) * ds + alpha / h * g * v * ds


# Interior penalty/DG terms
def dg_flux(u, v):
   return -ufl.dot(ufl.avg(ufl.grad(u)), ufl.jump(v, n))

F += dg_flux(u, v) * dS + dg_flux(v, u) * dS
F += gamma / ufl.avg(h) * ufl.inner(ufl.jump(v, n), ufl.jump(u, n)) * dS

a, L = ufl.system(F)
```
`script.py`
```python
from mpi4py import MPI
import dolfinx.fem.petsc
import numpy as np

from dg_form import a, L, alpha, gamma, el,f 


mesh = dolfinx.mesh.create_unit_square(MPI.COMM_WORLD, 6, 8, dolfinx.cpp.mesh.CellType.triangle)
Vh = dolfinx.fem.functionspace(mesh, el)



compiled_a = dolfinx.fem.compile_form(
    MPI.COMM_WORLD, a, form_compiler_options={"scalar_type": dolfinx.default_scalar_type}
)
compiled_L = dolfinx.fem.compile_form(
    MPI.COMM_WORLD, L, form_compiler_options={"scalar_type": dolfinx.default_scalar_type}
)


# Create coefficients and constants
alp = dolfinx.fem.Constant(mesh, 25.0)
gam = dolfinx.fem.Constant(mesh, 25.0)
fh = dolfinx.fem.Function(Vh)
fh.interpolate(lambda x: x[0] + 2*np.sin(x[1]))



a_form = dolfinx.fem.create_form(compiled_a, [Vh, Vh], mesh, {}, {alpha: alp, gamma: gam})
L_form = dolfinx.fem.create_form(compiled_L, [Vh], mesh, {f: fh}, {alpha: alp})

uh = dolfinx.fem.Function(Vh)
solver_options = {"ksp_type": "preonly", "pc_type": "lu", "pc_factor_mat_solver_type": "mumps"}
problem = dolfinx.fem.petsc.LinearProblem(a_form, L_form, u=uh, petsc_options=solver_options)
problem.solve()
print(f"Solver converged with {problem.solver.getConvergedReason()}")

with dolfinx.io.VTXWriter(mesh.comm, "solution.bp", [uh], engine="BP4") as bp:
    bp.write(0.0)    
```
With this split we can use the same `dg_form.py` in the C++ interface.

File: python/dolfinx/fem/forms.py [hunk 4]
```diff
@@ -96,6 +100,54 @@ def integral_types(self):
...
+def get_integration_domains(
+    integral_type: IntegralType,
+    subdomain: typing.Optional[typing.Union[MeshTags, list[tuple[int, np.ndarray]]]],
+    subdomain_ids: list[int],
+) -> list[tuple[int, np.ndarray]]:
+    """Get integration domains from subdomain data.
+
+    The subdomain data is a meshtags object consisting of markers, or a None object.
+    If it is a None object we do not pack any integration entities.
+    Integration domains is defined as a list of tuples, where each input `subdomain_ids`
```
Comment (IgorBaratta) [inline python/dolfinx/fem/forms.py:112]:
```suggestion
    Integration domains are defined as a list of tuples, where each input `subdomain_ids`
```

File: python/dolfinx/fem/forms.py [hunk 5]
```diff
@@ -333,3 +360,139 @@ def unique_spaces(V):
...
+
+def form_cpp_creator(
+    dtype: npt.DTypeLike,
+) -> typing.Union[
+    _cpp.fem.Form_float32,
+    _cpp.fem.Form_float64,
+    _cpp.fem.Form_complex64,
+    _cpp.fem.Form_complex128,
+]:
+    """Return the wrapped C++ class of a variational form of a specific scalar type.
```
Comment (IgorBaratta) [inline python/dolfinx/fem/forms.py:402]: class as in OO sense or a set/category? 
It seems to me that it's returning a creator/construtor function (as per usage below).

**Comment by chrisrichardson** [comment]
`form_compiler_options={"scalar_type": dolfinx.default_scalar_type}` - maybe this should be the default for `compile_form` ?

File: python/dolfinx/fem/forms.py (hunk 5 above)
Comment (jorgensd) [inline python/dolfinx/fem/forms.py:402]:
```suggestion
    """Return the wrapped C++ constructor for creating a variational form of a specific scalar type.
```

**Review by IgorBaratta — APPROVED** [review]
LGTM!

**Comment by jorgensd** [comment]
> `form_compiler_options={"scalar_type": dolfinx.default_scalar_type}` - maybe this should be the default for `compile_form` ?

We could do that, I have no strong opinions.

**Review by chrisrichardson — APPROVED** [review]

**Pull request merged**